src/django_components/component.py

```diff
@@ -989,28 +997,83 @@ def _render_impl(
 
         # Required for compatibility with Django's {% extends %} tag
         # See https://github.com/django-components/django-components/pull/859
-        context.render_context.push({BLOCK_CONTEXT_KEY: context.render_context.get(BLOCK_CONTEXT_KEY, {})})
+        context.render_context.push({BLOCK_CONTEXT_KEY: context.render_context.get(BLOCK_CONTEXT_KEY, BlockContext())})
 
         # By adding the current input to the stack, we temporarily allow users
         # to access the provided context, slots, etc. Also required so users can
         # call `self.inject()` from within `get_context_data()`.
+        #
+        # This is handled as a stack, as users can potentially call `component.render()`
+        # from within component hooks. Thus, then they do so, `component.id` will be the ID
+        # of the deepest-most call to `component.render()`.
         render_id = gen_id()
-        self._render_stack.append(
-            RenderStackItem(
-                input=RenderInput(
-                    id=render_id,
-                    context=context,
-                    slots=slots,
-                    args=args,
-                    kwargs=kwargs,
-                    type=type,
-                    render_dependencies=render_dependencies,
-                ),
-                is_filled=None,
+        metadata = MetadataItem(
+            render_id=render_id,
+            input=RenderInput(
+                context=context,
+                slots=slots,
+                args=args,
+                kwargs=kwargs,
+                type=type,
+                render_dependencies=render_dependencies,
             ),
+            is_filled=None,
         )
 
-        context_data = self.get_context_data(*args, **kwargs)
+        # If any slot fills were defined within the template, we want to scope them
+        # to the CSS of the parent component. Thus we keep track of the parent component.
+        if context.get(_COMPONENT_CONTEXT_KEY, None):
+            parent_id = cast(str, context[_COMPONENT_CONTEXT_KEY])
+            parent_comp_ctx = component_context_cache[parent_id]
+            component_path = [*parent_comp_ctx.component_path, self.name]
+        else:
+            parent_id = None
+            component_path = [self.name]
+
+        trace_component_msg(
+            "COMP_PREP_START",
+            component_name=self.name,
+            component_id=render_id,
+            slot_name=None,
+            component_path=component_path,
+            extra=f"Received {len(args)} args, {len(kwargs)} kwargs, {len(slots)} slots",
+        )
+
+        # Register the component to provide
+        register_provide_reference(context, render_id)
```

Comment:
Here we register that this component is a descendant of any provide data that it has access to, so that we can do refence counting.